linux/LinuxProcessList.c

```diff
@@ -704,6 +713,34 @@ static bool LinuxProcessList_readCmdlineFile(Process* process, const char* dirna
    process->basenameOffset = tokenEnd;
    setCommand(process, command, lastChar + 1);
 
+   LinuxProcess *lp = (LinuxProcess *)process;
+   lp->procCmdlineBasenameOffset = tokenStart;
+
+   /* /proc/[pid]/comm could change, so should be udpated */
+   free(lp->procComm);
+   xSnprintf(filename, MAX_NAME, "%s/%s/comm", dirname, name);
+   if ((fd = open(filename, O_RDONLY)) != -1 &&
+       (amtRead = xread(fd, command, sizeof(command) - 1)) > 0) {
+      close(fd);
+      command[amtRead - 1] = 0;
```

Comment:
Is the last char read always one we want to discard? Like `\n`?

Reply:
1. Regarding caching: LinuxProcessList_readCmdlineFile() indeed stores, in structure LinuxProcess: procComm, procExe, procExeBasenameOffset, procCmdlineBasenameOffset (argv0base in your comment); and in contained structure Process: cmdline (in a member calling comm - this is the preexisting name, though confusing in the present context). So is your question why not move the procComm etc. from LinuxProcess, to the Process structure? That is because the new members are Platform (Linux) dependent. But if the question is why we re-read the members, it is because comm, cmdline can be changed anytime by the process, and even exe is changed by kernel when the process execve's (even while the pid remains the same.)
2. The processing of cmdline is actually tricky. Normally processes store the arguments there separated by NUL (which are translated to '\n' when forming the cmdline string). But there are processes like chrome that store the entire argument vector separated by spaces (fun fact: in the upstream htop if you set hightlight basename, chrome's entire cmdline is highlighted!). Still other processes modify cmdline arbitrarily. In other words, the end of argv[0] is not readily/reliably obtainable from cmdline. So procCmdlineBasenameOffset in that argv[0] is also unreliable. I have left an example in the comments of matchCmdlinePrefixWithExeSuffix() "LinuxProcess.c:252" on how htop could get procCmdlineBasenameOffset wrong.
3. Regarding merging in matchCmdlinePrefixWithExeSuffix(): Case a) is simple enough - it applies when cmdline[0] == '/' (absolute path, since exe always begins with '/'). b) is just a special case of c with procCmdlineBasenameOffset == 0. c) is a special case of d) that follows: Due to the unreliability of procCmdlineBasenameOffset, upon a match failure we have to search the cmdline backwards to see if there is another potential procCmdlineBasenameOffset - this is the outer iteration. d) This is the general case we wish to handle - cmdline prefix (not just the basename) is a suffix of exe. For example exe == '/A1/B2/C3/E4' and cmdine = 'C3/E4 ARG1 ARG2' and we want merge to succeed. In this case we match E4 (procCmdlineBasenameOffset) and reverse match upto C3, upto a valid relative path in exe (i.e. cmdline == '3/E4' should not match). This is the inner iteration, and it covers c) as well.

> Is the last char read always one we want to discard? Like `\n`?

No this is just me attempting to be safe. It is true that when reading /proc/[pie]/comm we can expect that comm will be already NUL terminated.

Reply:
4. LinuxProcess_writeCommand() where comm is to matched first in exe's basename and then located in cmdline (if findCommInCmdline is set), is a bit complex mainly because of the various cases to be considered: coloring the three components while showing all three of them, merging two of them or merging all of them; showing or not showing full path, highlighting basename etc.,

Reply:
With caching of those three values I mostly referred to "per refresh cycle". Didn't also take your case d) (cmd: `C3/E4 args…`) into account. My questions were based on a much more focussed view on primarily highlighting the basename and trying to match based on that.

Possibly related is the display of `/snap/chromium/1298/usr/lib/chromium-browser/chrome` being completely marked as "basename" without that patch, although for some other chrome processes skipping of the path AND proper highlighting of the basename works correctly. Have to test again with this patch, if the issue persists.

Reply:
>Possibly related is the display of /snap/chromium/1298/usr/lib/chromium-browser/chrome being completely marked as "basename" without that patch, although for some other chrome processes skipping of the path AND proper highlighting of the basename works correctly. Have to test again with this patch, if the issue persists.

But do note that without the setting 'Merge exe, comm and cmdline in Command' (toggled by 'm' key), even this pull-request falls back to old behavior of htop of displaying cmdline - then you would see the issue. With the setting (and with exe readable) of course this issue will not be seen as then basename is that of the exe, and reliably known.

Reply:
Maybe have another PR for this after this is merged, that improves on extracting the correct basename in these cases. Based on some simple heuristic: If cmdline is `argv[0]` only assume separation by spaces instead. The exe is then assumed to be the first part before a space; from which basename is extracted.

Reply:
But identifying the basename on just space delimiter will mishandle cases where a directory name has spaces, like
cmdline == '/A1/B 2/E3 ARG1 ARG2'. Just saying there is no heuristics that can always correctly identify basename based upon cmdline alone, given the state of affairs that argv[0] is not always delimited by NUL in /proc/[pid]/cmdline and in fact can be modified arbitrarily. I do think the old basename identification heuristics in htop is reasonable (best effort possible) if it has only cmdline to work with. This is one of the reason for merging with the reliable exe (the other, main, reason being getting to know the path of the exe reliably).

Reply:
As it stands currently half of the processes started from chrome show chrome marked proerly as their basename and the others do not, which is kinda irritating.

Reply:
Are you referring to the the upstream htop that uses only cmdline for Command column (same as the unmerged Command that uses only cmdline in this pull request)? Yes, this is a known limitation with only using cmdline that I had already described. But the merged Command (toggle using 'm') in this pull request does fix this issue properly by using exe for basename - as I said this is one of the motivations of this pull request. Please see chrome in below screenshot of this pull request, with the setting "Merge exe, comm and cmdline in Command" ('m' toggle). However, please clarify if you mean some other issue with this pull request itself.
![Screenshot from 2020-09-28 09-57-29](https://user-images.githubusercontent.com/23692893/94390719-10d10c80-0171-11eb-936c-64756a2b39dd.png)